[FLINK-14033][yarn] upload user artifacts for yarn job cluster

HuangZhenQiu opened the pull request:
## What is the purpose of the change

Upload user artifacts that registered in distributed caches for yarn job clusters. Currently, distributed caches are not handled correctly. The solution proposed is to upload local files onto the remote file system and update the registered file in cache entry. 

## Brief change-log
  - Modified the AbstractClusterDescriptor to handle with artifacts for per job mode before starting 
     application master.

## Verifying this change

The change is end to end testing in YarnDistributedCacheITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -738,6 +738,25 @@ public ApplicationReport startAppMaster(
+				.map(entry -> Tuple2.of(entry.getKey(), new org.apache.flink.core.fs.Path(entry.getValue().filePath)))
+				.collect(Collectors.toList());
+
+			for (Tuple2<String, org.apache.flink.core.fs.Path> userArtifact : userArtifacts) {
```
Can we use `for (Map.Entry<String, DistributedCache.DistributedCacheEntry> entry : jobGraph.getUserArtifacts().entrySet())` for the loop here? We may not need to create the userArtifacts list.

HuangZhenQiu replied:
@wangyang0918 
I updated the loop and added an integration test in yarn test module. Would you please take one more round of review?

HuangZhenQiu commented:
@wangyang0918 @tillrohrmann 
Would you please help to review the PR. One of our customers wants to use this feature for his job.

HuangZhenQiu commented:
@wangyang0918 @suez1224 @tillrohrmann I just rebased lastest master, would you please take one more round of review?

HuangZhenQiu commented:
@aljoscha Given @tillrohrmann Is on vacation, would you please help to review this PR?

aljoscha commented:
Thanks for this contribution, I think it is a valuable feature! Could you please rebase this on top of the latest master and clean up the commits? I noticed some dangling/unused changes, for example you add a `CacheBasedDataSource` that is not used, and in `YarnTestCacheJob` the mapper that does the actual testing is not used, meaning the test does not test what it is supposed to test.

Did you think about how the uploaded files are being cleaned up? What is the lifecycle? Will they stay in the DFS after the application finishes?

HuangZhenQiu commented:
@aljoscha Thanks for pointing out issues in test case. I rebased latest master and removed unused classes. The distributed cache functionality should be covered in the map function.  For the life cycle of the distributed cache files, we should split into two parts. For the local copies in each TM, they are managed as a rolling array in the FileCache class. Any time one acces to a file not in FileCache, the file will be read from remote. For the remote file, it is the same as other local resources files for yarn application, they are under the ./flink/application_id fodler. If user kill the job, the folder will be deleted.

HuangZhenQiu commented:
@aljoscha During debugging, I found another bug introduced by a recent change. https://issues.apache.org/jira/browse/FLINK-14982. I would like to fix it, please assign to me after confirming it is a real issue.

aljoscha commented:
Thanks! This looks good now. I'm currently building some things and debugging, but once I'm done with that I'll review this again and then merge. 👌

sunjincheng121 commented:
Just a reminder: This PR is the blocker of FLINK-14019 for YARN pre job mode. :)

HuangZhenQiu commented:
@sunjincheng121 Would you please also help to review the PR.

aljoscha commented:
Thanks a lot for this work! I merged it. 🍵
